File: _aoti_pt2/_cpp_src/test_exported_pt2.cpp
/*
 * This script has been kept independent of the F3 package
 * to make it easier to test the exported .pt2 files on Orin.
 * This way, there is no need to install the F3 package on Orin,
 * which has many dependencies and can be cumbersome to install.
 *
 * Author: Richeek Das
 */

#include <chrono>
#include <cstdint>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <string>

#include <torch/csrc/inductor/aoti_package/model_package_loader.h>
#include <torch/torch.h>

struct Config {
  std::string f3_pt2_path;
  std::string dav2_pt2_path;
  std::string flowhead_pt2_path;
  int32_t warmup_runs = 20;
  int32_t runs = 200;
  int32_t n_events = 200000;
  int32_t dav2_height = 238;
  int32_t dav2_width = 308;
  int32_t flow_height = 238;
  int32_t flow_width = 308;
};

void print_usage(const char *program_name) {
  std::cout
      << "Usage: " << program_name << " [OPTIONS]\n"
      << "\nRequired options:\n"
      << "  --f3_pt2_path PATH          Path to the F3 pt2 file\n"
      << "\nOptional options:\n"
      << "  --dav2_pt2_path PATH        Path to the DAV2 pt2 file\n"
      << "  --flowhead_pt2_path PATH    Path to the FlowHead pt2 file\n"
      << "  --runs N                    Number of timed inference runs "
         "(default: 200)\n"
      << "  --warmup_runs N             Number of warmup runs before timing "
         "(default: 20)\n"
      << "  --dav2_height H             Target height for DAV2 input (default: "
         "238)\n"
      << "  --dav2_width W              Target width for DAV2 input (default: "
         "308)\n"
      << "  --flow_height H             Target height for FlowHead input "
         "(default: 238)\n"
      << "  --flow_width W              Target width for FlowHead input "
         "(default: 308)\n"
      << "  --help                      Display this help message\n";
}

bool parse_args(int argc, char *argv[], Config &config) {
  bool f3_path_provided = false;

  for (int i = 1; i < argc; i++) {
    std::string arg = argv[i];

    if (arg == "--help" || arg == "-h") {
      print_usage(argv[0]);
      return false;
    } else if (arg == "--f3_pt2_path" && i + 1 < argc) {
      config.f3_pt2_path = argv[++i];
      f3_path_provided = true;
    } else if (arg == "--dav2_pt2_path" && i + 1 < argc) {
      config.dav2_pt2_path = argv[++i];
    } else if (arg == "--flowhead_pt2_path" && i + 1 < argc) {
      config.flowhead_pt2_path = argv[++i];
    } else if (arg == "--runs" && i + 1 < argc) {
      config.runs = std::stoi(argv[++i]);
    } else if (arg == "--warmup_runs" && i + 1 < argc) {
      config.warmup_runs = std::stoi(argv[++i]);
    } else if (arg == "--dav2_height" && i + 1 < argc) {
      config.dav2_height = std::stoi(argv[++i]);
    } else if (arg == "--dav2_width" && i + 1 < argc) {
      config.dav2_width = std::stoi(argv[++i]);
    } else if (arg == "--flow_height" && i + 1 < argc) {
      config.flow_height = std::stoi(argv[++i]);
    } else if (arg == "--flow_width" && i + 1 < argc) {
      config.flow_width = std::stoi(argv[++i]);
    } else {
      std::cerr << "Unknown or incomplete argument: " << arg << std::endl;
      print_usage(argv[0]);
      return false;
    }
  }

  if (!f3_path_provided) {
    std::cerr << "Error: --f3_pt2_path is required\n" << std::endl;
    print_usage(argv[0]);
    return false;
  }

  return true;
}

void print_config(const Config &config) {
  std::cout << "⚙️ Configuration:" << std::endl;
  std::cout << "    F3 PT2 path: " << config.f3_pt2_path << std::endl;
  std::cout << "    DAV2 PT2 path: "
            << (config.dav2_pt2_path.empty() ? "None" : config.dav2_pt2_path)
            << std::endl;
  std::cout << "    FlowHead PT2 path: "
            << (config.flowhead_pt2_path.empty() ? "None"
                                                 : config.flowhead_pt2_path)
            << std::endl;
  std::cout << "    DAV2 input shape: " << config.dav2_height << "x"
            << config.dav2_width << std::endl;
  std::cout << "    FlowHead input shape: " << config.flow_height << "x"
            << config.flow_width << std::endl;
  std::cout << "    Warmup runs: " << config.warmup_runs << std::endl;
  std::cout << "    Timed runs: " << config.runs << std::endl;
  std::cout << std::endl;
}

torch::Tensor generate_random_events(int32_t n_events) {
  // Generate random events (N, 4) with values in [0, 1) range
  // Columns: [x, y, t, p]
  return torch::rand({n_events, 4},
                     at::TensorOptions().dtype(at::kFloat).device(at::kCUDA)) *
         0.9f;
}

int main(int argc, char *argv[]) {
  c10::InferenceMode mode;

  Config config;

  if (!parse_args(argc, argv, config)) {
    return 1;
  }

  print_config(config);

  std::cout << "Loading F3 model..." << std::endl;
  auto start_load = std::chrono::high_resolution_clock::now();
  torch::inductor::AOTIModelPackageLoader f3_loader(config.f3_pt2_path);
  auto end_load = std::chrono::high_resolution_clock::now();

  auto load_duration = std::chrono::duration_cast<std::chrono::milliseconds>(
      end_load - start_load);
  std::cout << "F3 model loading time: " << load_duration.count() << " ms"
            << std::endl;

  // Load depth and flow models if paths are provided
  std::optional<torch::inductor::AOTIModelPackageLoader> dav2_loader;
  std::optional<torch::inductor::AOTIModelPackageLoader> flowhead_loader;

  if (!config.dav2_pt2_path.empty()) {
    std::cout << "Loading DAV2 model..." << std::endl;
    start_load = std::chrono::high_resolution_clock::now();
    dav2_loader.emplace(config.dav2_pt2_path);
    end_load = std::chrono::high_resolution_clock::now();
    load_duration = std::chrono::duration_cast<std::chrono::milliseconds>(
        end_load - start_load);
    std::cout << "DAV2 model loading time: " << load_duration.count() << " ms"
              << std::endl;
  }

  if (!config.flowhead_pt2_path.empty()) {
    std::cout << "Loading FlowHead model..." << std::endl;
    start_load = std::chrono::high_resolution_clock::now();
    flowhead_loader.emplace(config.flowhead_pt2_path);
    end_load = std::chrono::high_resolution_clock::now();
    load_duration = std::chrono::duration_cast<std::chrono::milliseconds>(
        end_load - start_load);
    std::cout << "FlowHead model loading time: " << load_duration.count()
              << " ms" << std::endl;
  }

  // For now, using random events
  std::vector<torch::Tensor> inputs = {generate_random_events(config.n_events)};

  std::cout << "\nInput events shape: [" << inputs[0].size(0) << ", "
            << inputs[0].size(1) << "]" << std::endl;

  // Warmup runs
  std::cout << "\nRunning warmup iterations..." << std::endl;
  for (int i = 0; i < config.warmup_runs; i++) {
    std::vector<torch::Tensor> f3_outputs = f3_loader.run(inputs);
    torch::Tensor f3_feat = f3_outputs[0].permute({0, 3, 2, 1});

    if (i == 0)
      std::cout << "F3 output shape: [" << f3_feat.size(0) << ", "
                << f3_feat.size(1) << ", " << f3_feat.size(2) << ", "
                << f3_feat.size(3) << "]" << std::endl;

    if (dav2_loader) {
      torch::Tensor f3_feat_ds = torch::nn::functional::interpolate(
          f3_feat,
          torch::nn::functional::InterpolateFuncOptions()
              .size(std::vector<int64_t>{config.dav2_height, config.dav2_width})
              .mode(torch::kBilinear)
              .align_corners(true));

      std::vector<torch::Tensor> dav2_inputs = {f3_feat_ds.to(at::kFloat)};

      std::vector<torch::Tensor> depth_outputs =
          dav2_loader.value().run(dav2_inputs);

      if (i == 0)
        std::cout << "Depth output shape: [" << depth_outputs[0].size(0) << ", "
                  << depth_outputs[0].size(1) << ", "
                  << depth_outputs[0].size(2) << "]" << std::endl;
    }

    if (flowhead_loader) {
      torch::Tensor f3_feat_flow = torch::nn::functional::interpolate(
          f3_feat,
          torch::nn::functional::InterpolateFuncOptions()
              .size(std::vector<int64_t>{config.flow_height, config.flow_width})
              .mode(torch::kBilinear)
              .align_corners(true));
      std::vector<torch::Tensor> flow_inputs = {f3_feat_flow.to(at::kFloat)};
      std::vector<torch::Tensor> flow_outputs =
          flowhead_loader.value().run(flow_inputs);

      if (i == 0)
        std::cout << "Flow output shape: [" << flow_outputs[0].size(0) << ", "
                  << flow_outputs[0].size(1) << ", " << flow_outputs[0].size(2)
                  << ", " << flow_outputs[0].size(3) << "]" << std::endl;
    }
  }

  torch::cuda::synchronize();

  // Timed runs
  std::cout << "\nRunning timed iterations..." << std::endl;
  auto start_inference = std::chrono::high_resolution_clock::now();
  for (int i = 0; i < config.runs; i++) {
    std::vector<torch::Tensor> f3_outputs = f3_loader.run(inputs);
    torch::Tensor f3_feat = f3_outputs[0].permute({0, 3, 2, 1});

    if (dav2_loader) {
      torch::Tensor f3_feat_ds = torch::nn::functional::interpolate(
          f3_feat,
          torch::nn::functional::InterpolateFuncOptions()
              .size(std::vector<int64_t>{config.dav2_height, config.dav2_width})
              .mode(torch::kBilinear)
              .align_corners(true));
      std::vector<torch::Tensor> dav2_inputs = {f3_feat_ds.to(at::kFloat)};
      std::vector<torch::Tensor> depth_outputs =
          dav2_loader.value().run(dav2_inputs);
    }

    if (flowhead_loader) {
      torch::Tensor f3_feat_flow = torch::nn::functional::interpolate(
          f3_feat,
          torch::nn::functional::InterpolateFuncOptions()
              .size(std::vector<int64_t>{config.flow_height, config.flow_width})
              .mode(torch::kBilinear)
              .align_corners(true));
      std::vector<torch::Tensor> flow_inputs = {f3_feat_flow.to(at::kFloat)};
      std::vector<torch::Tensor> flow_outputs =
          flowhead_loader.value().run(flow_inputs);
    }
  }
  torch::cuda::synchronize();
  auto end_inference = std::chrono::high_resolution_clock::now();

  auto total_duration = std::chrono::duration_cast<std::chrono::microseconds>(
      end_inference - start_inference);
  double avg_time_ms = (total_duration.count() / 1000.0) / config.runs;

  std::cout << "\nAverage inference time over " << config.runs
            << " runs: " << std::fixed << std::setprecision(2) << avg_time_ms
            << " ms" << std::endl;

  return 0;
}

File: _aoti_pt2/_py_src/test_exported_pt2.py
"""
This script has been kep independent of the f3 package
to make it easier to test the exported pt2 files on orin.
This way one does not need to install the f3 package on orin,
which has many dependencies and is a bit cumbersome to install.

Author: Richeek Das
"""

import cv2
import h5py
import time
import argparse
import numpy as np
from sklearn.decomposition import PCA

import torch
import torch._inductor
import torch.nn.functional as F

torch.set_float32_matmul_precision('high')


def plot_events(events: torch.Tensor, img_width: int, img_height: int):
    """Plot events onto an image and save to disk."""
    img = torch.zeros((img_height, img_width, 3), device=events.device)
    coords_x = (events[:, 0] * img_width).round().int()
    coords_y = (events[:, 1] * img_height).round().int()
    polarities = events[:, 3]
    img[coords_y, coords_x, 0] += (polarities == 1).float()
    img[coords_y, coords_x, 2] += (polarities == 0).float()
    img = torch.clamp(img, 0, 1)
    if img.device.type == 'cuda':
        img = img.cpu()
    img = (img.numpy() * 255).astype("uint8")
    cv2.imwrite("events_plot.png", img)
    return img


def plot_patched_features(feat):
    """
    Args:
        feat: (W, H, LF) torch.Tensor or np.ndarray
    """
    W, H, LF = feat.shape

    if isinstance(feat, torch.Tensor):
        feat = feat.cpu().numpy()

    feat = feat.reshape(-1, LF)
    # PCA to reduce the dimensionality of the last dimension
    pca = PCA(n_components=min(3, LF))
    pca_result = pca.fit_transform(feat)
    mean, std = pca_result.mean(axis=0), pca_result.std(axis=0)
    pca_result = np.clip(pca_result, mean - std, mean + std)
    pca_result = (pca_result - pca_result.min()) / (pca_result.max() - pca_result.min())
    pca_result = (pca_result * 255).astype(np.uint8).reshape(W, H, -1)
    cv2.imwrite("f3_feat_pca.png", pca_result)


def plot_depth(depth):
    """
    Plot depth map with proper normalization and save to disk.
    
    Args:
        depth: torch.Tensor or np.ndarray, depth map to visualize
        output_filename: str, filename to save the visualization
    """
    # Postprocessing, implement this part in your chosen language:
    # cmap = cv2.applyColorMap(np.arange(0, 256, dtype=np.uint8), cv2.COLORMAP_JET)
    if isinstance(depth, torch.Tensor):
        depth = depth.cpu().numpy()
    
    if depth.ndim == 3:
        depth = depth[0]

    mask = ~np.isnan(depth)
    min_depth = depth[mask].min()
    max_depth = depth[mask].max()
    depth_colored = np.clip(depth, min_depth, max_depth)
    depth_colored = (depth_colored - min_depth) / (max_depth - min_depth)
    depth_colored[~mask] = 0
    depth_colored = (depth_colored * 255).astype(np.uint8)
    cv2.imwrite("depth_colored.png", depth_colored)


def flow_viz_np(flow, norm=False):
    h, w = flow.shape[:2]
    mag, ang = cv2.cartToPolar(flow[..., 0], flow[..., 1])
    hsv = np.zeros((h, w, 3), dtype=np.uint8)
    hsv[:, :, 0] = ang * 180 / np.pi / 2
    hsv[:, :, 1] = 255
    if norm:
        magmean, magstd = np.mean(mag), np.std(mag)
        mag = np.clip(mag, magmean - 2 * magstd, magmean + 2 * magstd)
    hsv[:, :, 2] = cv2.normalize(mag, None, 0, 255, cv2.NORM_MINMAX)
    flow_rgb = cv2.cvtColor(hsv, cv2.COLOR_HSV2BGR)
    return flow_rgb


def plot_flow(flow, event_img=None):
    """
    Plot optical flow visualization and save to disk.
    
    Args:
        flow: torch.Tensor or np.ndarray of shape (B, 2, H, W) or (2, H, W) or (H, W, 2)
    """
    if isinstance(flow, torch.Tensor):
        if flow.ndim == 4:  # (B, 2, H, W)
            flow = flow[0]  # Take first batch
        flow = flow.cpu().numpy()

    flow_rgb = flow_viz_np(flow.transpose(1, 2, 0) if flow.shape[0] == 2 else flow)

    if event_img is not None:
        overlay_image = flow_rgb.copy()
        mask = (event_img == 255).astype(np.uint8).max(axis=-1)
        # interpolate mask to match overlay_image size if needed
        if overlay_image.shape[0] != mask.shape[0] or overlay_image.shape[1] != mask.shape[1]:
            mask = cv2.resize(mask, (overlay_image.shape[1], overlay_image.shape[0]), interpolation=cv2.INTER_NEAREST)
        overlay_image *= mask[..., None]
        cv2.imwrite("flow_viz_overlay.png", overlay_image)

    cv2.imwrite("flow_viz.png", flow_rgb)


def load_events_from_h5(h5_file_path, time_ms,
                        img_width=1280, img_height=720, time_window_ms=20):
    """
    Load events from M3ED H5 file and return normalized events tensor.
    
    Args:
        h5_file_path: str, path to the H5 file
        start_time_ms: int, start time in milliseconds
        n_events: int, maximum number of events to load
        img_width: int, image width for normalization (default: 1280)
        img_height: int, image height for normalization (default: 720)
        time_window_ms: int, time window in milliseconds (default: 20)
    
    Returns:
        torch.Tensor: normalized events tensor of shape (N, 4) on CUDA
    """
    hdf5_file = h5py.File(h5_file_path, 'r')
    events_x = hdf5_file['prophesee/left/x']
    events_y = hdf5_file['prophesee/left/y']
    events_t = hdf5_file['prophesee/left/t']
    events_p = hdf5_file['prophesee/left/p']
    ms_to_idx = hdf5_file['prophesee/left/ms_map_idx']

    norm_factor = torch.tensor([
        img_width, img_height, time_window_ms, 1
    ], dtype=torch.float32)

    ei = ms_to_idx[time_ms]
    si = ms_to_idx[time_ms - time_window_ms]

    events_input = torch.zeros((ei - si, 4), dtype=torch.float32)
    events_input[:, 0] = torch.from_numpy(events_x[si:ei])
    events_input[:, 1] = torch.from_numpy(events_y[si:ei])
    events_input[:, 2] = time_ms - torch.from_numpy(events_t[si:ei]) // 1000 - 1
    events_input[:, 3] = torch.from_numpy(events_p[si:ei])

    events_input = events_input / norm_factor

    events_input = events_input.cuda()

    return events_input


def count_parameters(model):
    """Count total parameters in a model"""
    return sum(p.numel() for p in model.parameters())


def compile_engine_and_infer(f3_pt2_path, dav2_pt2_path, flowhead_pt2_path, h5_file_path, time_ms, 
                             runs, warmup_runs, dav2_height, dav2_width, flow_height, flow_width):
    # Load events from H5 file
    ctx = load_events_from_h5(h5_file_path, time_ms)
    event_img = plot_events(ctx, 1280, 720)

    aoti_f3_compiled = torch._inductor.aoti_load_package(f3_pt2_path)
    if dav2_pt2_path is not None:
        aoti_dav2_compiled = torch._inductor.aoti_load_package(dav2_pt2_path)
    if flowhead_pt2_path is not None:
        aoti_flowhead_compiled = torch._inductor.aoti_load_package(flowhead_pt2_path)

    # Warm-up runs
    for _ in range(warmup_runs):
        f3_feat = aoti_f3_compiled(ctx).permute(0, 3, 2, 1)
        if dav2_pt2_path is not None:
            f3_feat_ds = F.interpolate(
                f3_feat,
                size=(dav2_height, dav2_width),
                mode='bilinear',
                align_corners=True
            )
            depth = aoti_dav2_compiled(f3_feat_ds.float())
        if flowhead_pt2_path is not None:
            f3_feat_flow = F.interpolate(
                f3_feat,
                size=(flow_height, flow_width),
                mode='bilinear',
                align_corners=True
            )
            flow = aoti_flowhead_compiled(f3_feat_flow.float())

    if dav2_pt2_path is not None:
        print(f"Depth output shape: {depth.shape}")
    if flowhead_pt2_path is not None:
        print(f"Flow output shape: {flow.shape}")

    plot_patched_features(f3_feat[0].permute(1, 2, 0))

    if dav2_pt2_path is not None:
        plot_depth(depth.float())
    if flowhead_pt2_path is not None:
        plot_flow(flow.float(), event_img=event_img)

    # Timed runs
    torch.cuda.synchronize()
    start_time = time.time()
    for _ in range(runs):
        f3_feat = aoti_f3_compiled(ctx).permute(0, 3, 2, 1)
        if dav2_pt2_path is not None:
            f3_feat_ds = F.interpolate(
                f3_feat,
                size=(dav2_height, dav2_width),
                mode='bilinear',
                align_corners=True,
            )
            depth = aoti_dav2_compiled(f3_feat_ds.float())
        if flowhead_pt2_path is not None:
            f3_feat_flow = F.interpolate(
                f3_feat,
                size=(flow_height, flow_width),
                mode='bilinear',
                align_corners=True,
            )
            flow = aoti_flowhead_compiled(f3_feat_flow.float())
    torch.cuda.synchronize()
    end_time = time.time()

    total_time = end_time - start_time
    print(f"Average inference time over {runs} runs: {total_time / runs * 1000:.2f} ms")

    torch._dynamo.reset()


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='AOTI inference test')
    parser.add_argument('--f3_pt2_path', type=str, required=True,
                        help='Path to the F3 pt2 file')
    parser.add_argument('--dav2_pt2_path', type=str, default=None,
                        help='Path to the DAV2 pt2 file')
    parser.add_argument('--flowhead_pt2_path', type=str, default=None,
                        help='Path to the FlowHead pt2 file')
    parser.add_argument('--h5_file', type=str, default='car_urban_day_penno_small_loop_data.h5',
                        help='Path to the M3ED H5 dataset file (default: car_urban_day_penno_small_loop_data.h5)')
    parser.add_argument('--time_ms', type=int, default=30000,
                        help='Time in milliseconds for event extraction (default: 30000)')
    parser.add_argument('--runs', type=int, default=200,
                        help='Number of timed inference runs (default: 200)')
    parser.add_argument('--warmup_runs', type=int, default=20,
                        help='Number of warmup runs before timing (default: 20)')
    parser.add_argument('--dav2_height', type=int, default=238,
                        help='Target height for DAV2 input (default: 238)')
    parser.add_argument('--dav2_width', type=int, default=308,
                        help='Target width for DAV2 input (default: 308)')
    parser.add_argument('--flow_height', type=int, default=238,
                        help='Target height for FlowHead input (default: 238)')
    parser.add_argument('--flow_width', type=int, default=308,
                        help='Target width for FlowHead input (default: 308)')

    args = parser.parse_args()

    print("⚙️ Configuration:")
    print(f"    H5 file: {args.h5_file}")
    print(f"    Start time: {args.time_ms} ms")
    print(f"    F3 PT2 path: {args.f3_pt2_path}")
    print(f"    DAV2 PT2 path: {args.dav2_pt2_path if args.dav2_pt2_path else 'None'}")
    print(f"    FlowHead PT2 path: {args.flowhead_pt2_path if args.flowhead_pt2_path else 'None'}")
    print(f"    DAV2 input shape: {args.dav2_height}x{args.dav2_width}")
    print(f"    FlowHead input shape: {args.flow_height}x{args.flow_width}")
    print(f"    Warmup runs: {args.warmup_runs}")
    print(f"    Timed runs: {args.runs}")
    print()

    compile_engine_and_infer(
        args.f3_pt2_path,
        args.dav2_pt2_path,
        args.flowhead_pt2_path,
        args.h5_file,
        args.time_ms,
        args.runs,
        args.warmup_runs,
        args.dav2_height,
        args.dav2_width,
        args.flow_height,
        args.flow_width
    )
